### ISSUE TITLE
Possible race condition when calling custom op with control dependencies

### ISSUE THREAD

**Issue by chandanjc** [tensorflow/tensorflow#29461]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v1.12.0-0-ga6d8ffae09 1.12.0
- Python version: 3.5
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I was trying to study the behavior of tf.control_dependencies when training Resnet50 on distributed CPU, with the benchmark script (https://github.com/tensorflow/benchmarks/tree/cnn_tf_v1.12_compatible/scripts/tf_cnn_benchmarks). 
I have modified this script to load the same dataset on each run, and removed the random params from the model (for predictable repro). 
I have written a dummy custom C++ op where I'm simply printing the input tensor and returning.
I have also modified the benchmark_cnn.py in the above repo to call my custom op, instead of hvd.allreduce. I have also added control dependency between the layers to ensure that my custom op gets called one layer after another. Following are the code changes for this:

```
        pred = None
        for i in range(0, len(grads), 1):
            grad = grads[i]

            import re
            if re.search("batch.*norm", params[i].name, re.IGNORECASE) != None:
              grads[i] = grad
            else:
              tensor_name = 'HorovodAllreduce_%s' % hvd_normalize_name(grad.name)
              if pred != None:
                with tf.control_dependencies([pred]):
                  grad = custom_op_cpp(grad, horovod_device, tensor_name, hvd.rank())
              else:
                grad = custom_op_cpp(grad, horovod_device, tensor_name, hvd.rank())

              pred = grad
              grads[i] = grad
```

And custom_op_cpp() function is defined as follows:
```
custom_module = tf.load_op_library('./custom_op_cpp/custom_op.so')

def custom_op_cpp(t_data, dense_device_name, tensor_name, my_rank):
    result = custom_module.dummy_func(t_data, tensor_name, my_rank)
    return result
```

Custom C++ op is defined as:
```
REGISTER_OP("DummyFunc")
    .Input("input: float32")
    .Input("tname: string")
    .Input("rank: int32")
    .Output("output: float32")
    .SetShapeFn([](::tensorflow::shape_inference::InferenceContext* c) {
        c->set_output(0, c->input(0));
        return Status::OK();
    });

class DummyFuncOp : public OpKernel {
    explicit DummyFuncOp(OpKernelConstruction* context) : OpKernel(context) {}

    void Compute(OpKernelContext* context) override {
        ....
        const Tensor& input_tensor = context->input(0);
        float *inp_tensor = (float *)(input_tensor.tensor_data().data());
        size_t inp_tensor_size = (size_t)input_tensor.shape().num_elements();
          
        PrintTensorData(inp_tensor, inp_tensor_size);

       //Some code to return copy the input tensor into output and return
    }
}
```

This works as expected, I see the same gradients for each run (1 iteration), since I'm loading the same dataset each time. 
Now, If I reverse the control dependencies from the benchmark script as follows:
`for i in reversed(range(0, len(grads), 1)):`
Then I see that the gradients coming into the dummy op are different for some of the layers, and they are random from run to run. However, if I add a `sleep(4)` in my dummy op's Compute function, then this issue is no longer seen. Hence, I'm suspecting that there may be a race condition, wherein dummy op is getting invoked even before the gradients are properly updated in the backprop phase. 

**Describe the expected behavior**
Whether default or reverse control dependency is used between layers should not have any implications on the gradients seen in dummy c++ op for a given layer, since we are running a predictable model, with same datasets. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
@chandanjc In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by chandanjc** [comment]
@gadagashwini Do you need me to share the code for c++ op or all of the Resnet50 benchmark script changes to make it deterministic?

**Comment by gadagashwini** [comment]
@chandanjc Will it be possible to provide us the minimal full code snippet that includes all your operations. So that we can reproduce it on our environment. Thanks!

**Comment by gadagashwini** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
